### PR TITLE
many: let provide a SnapHandler to Seed.Load*Meta*

### PR DESCRIPTION
**Pull request by pedronis**
This allows to perform dedicated handling for seed snaps (like copying
them already) together with digest computation.

SnapHandler is slightly complex because of unasserted vs asserted
snaps and the differences how UC16/18 vs 20+ seeds are processed.

Before there was always caching of essential snaps across
LoadEssentialMeta* and LoadMeta, this is turned off if handlers are
provided, OTOH specific caching can be done by handler logic if it makes sense.

### CODE REVIEW

File: seed/seed16.go [hunk 1]
```diff
@@ -350,10 +377,17 @@ func (s *seed16) LoadEssentialMeta(essentialTypes []snap.Type, tm timings.Measur
 
 	s.resetSnaps()
 
-	return s.loadEssentialMeta(essentialTypes, required, added, tm)
+	if handler == nil {
+		handler = defaultSnapHandler{}
+	}
+
+	if len(essentialTypes) == 0 {
```
Comment (pedronis) [inline seed/seed16.go:384]: this is a slightly unrelated change because I caught that we had changed the seed20 behavior about this but not the seed16, see also related test change below

File: seed/seed16.go [hunk 2]
```diff
@@ -147,12 +149,33 @@ func (s *seed16) addSnap(sn *internal.Snap16, pinnedTrack string, cache map[stri
...
+				if err != nil {
+					return snap.Revision{}, err
+				}
+				return si.Revision, err
+			}
 			timings.Run(tm, "derive-side-info", fmt.Sprintf("hash and derive side info for snap %q", sn.Name), func(nested timings.Measurer) {
-				si, err = snapasserts.DeriveSideInfo(path, s.db)
+				var snapSHA3_384 string
+				var snapSize uint64
+				snapSHA3_384, snapSize, err = handler.HandleAndDigestAssertedSnap(sn.Name, path, essType, nil, deriveRev, tm)
```
Comment (mardy) [inline seed/seed16.go:173]: Wouldn't this work as well?
```suggestion
				snapSHA3_384, snapSize, err := handler.HandleAndDigestAssertedSnap(sn.Name, path, essType, nil, deriveRev, tm)
```

**Review by mardy — APPROVED** [review]
LGTM, thanks!

File: seed/seed16.go [hunk 3]
```diff
@@ -124,11 +124,13 @@ func (s *seed16) SetParallelism(int) {
...
 
-func (s *seed16) addSnap(sn *internal.Snap16, pinnedTrack string, cache map[string]*Snap, tm timings.Measurer) (*Snap, error) {
+func (s *seed16) addSnap(sn *internal.Snap16, essType snap.Type, pinnedTrack string, handler SnapHandler, cache map[string]*Snap, tm timings.Measurer) (*Snap, error) {
 	path := filepath.Join(s.seedDir, "snaps", sn.File)
 
+	_, defaultHandler := handler.(defaultSnapHandler)
+
 	seedSnap := cache[path]
-	if seedSnap == nil {
+	if seedSnap == nil || !defaultHandler {
```
Comment (stolowski) [inline seed/seed16.go:133]: The `!defaultHandler` condition is not clear to me, can you add a comment in the code?

File: spread.yaml [hunk 4]
```diff
@@ -44,7 +44,7 @@ environment:
     UBUNTU_IMAGE_SNAP_CHANNEL: "latest/candidate"
     # controls whether ubuntu-image is built using the current snapd tree as a
     # dependency or the one listed in its go.mod
-    UBUNTU_IMAGE_ALLOW_API_BREAK: '$(HOST: echo "${SPREAD_UBUNTU_IMAGE_ALLOW_API_BREAK:-true}")'
+    UBUNTU_IMAGE_ALLOW_API_BREAK: '$(HOST: echo "${SPREAD_UBUNTU_IMAGE_ALLOW_API_BREAK:-false}")'
```
Comment (stolowski) [inline spread.yaml:47]: What's the significance of this change?

File: seed/seed20.go [hunk 5]
```diff
@@ -331,14 +331,17 @@ func (s *seed20) lookupSnap(snapRef naming.SnapRef, optSnap *internal.Snap20, ch
 			return nil, fmt.Errorf("cannot read unasserted snap: %v", err)
 		}
 		sideInfo = &snap.SideInfo{RealName: info.SnapName()}
+		if err := handler.HandleUnassertedSnap(info.SnapName(), path, tm); err != nil {
```
Comment (stolowski) [inline seed/seed20.go:334]: Nitpick, sideInfo could be assigned afterwards (if err == nil)?

**Review by stolowski — APPROVED** [review]
Thanks, looks nice, a few minor questions.

File: spread.yaml (hunk 4 above)
Comment (pedronis) [reply spread.yaml:47]: ubuntu-image needs changes because it uses LoadMeta, so until it's fixed we need to use this to stop building it together with snapd from the branch

File: seed/seed16.go (hunk 3 above)
Comment (pedronis) [reply seed/seed16.go:133]: is to ignore the cache if a handler is passed in, as explained in the doc comment of the methods. If we use the cache we wouldn't call the handler for the cached snap which could be unexpected. I'll add comments in code

File: seed/seed16.go (hunk 2 above)
Comment (mardy) [reply seed/seed16.go:173]: Ah, right, thanks for the explanation!

File: seed/seed16.go [hunk 6]
```diff
@@ -147,12 +149,33 @@ func (s *seed16) addSnap(sn *internal.Snap16, pinnedTrack string, cache map[stri
...
 			var err error
+
+			deriveRev := func(snapSHA3_384 string, snapSize uint64) (snap.Revision, error) {
+				var err error
+				if si == nil {
+					si, err = snapasserts.DeriveSideInfoFromDigestAndSize(path, snapSHA3_384, snapSize, s.db)
+				}
+				if err != nil {
+					return snap.Revision{}, err
+				}
```
[the remark this answers is not in the document]
Comment (pedronis) [reply seed/seed16.go:167]: it can be moved but the declaration is needed because we don't want to shadow the outside si
